include/Keymap.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Keymap.h - holds information about a key mapping
+ *
+ * Copyright (c) 2020 Martin Pavelek <he29.HS/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef KEYMAP_H
+#define KEYMAP_H
+
+#include <vector>
+#include <QObject>
+#include <QString>
+
+#include "Note.h"
+#include "SerializingObject.h"
+
+class Keymap : public QObject, public SerializingObject
+{
+	Q_OBJECT
+public:
+	Keymap();
+	Keymap(	QString description,
+			std::vector<int> newMap,
```

Comment:
In the place where the constructor is called, the vector is "marked" with `std::move`; wouldn't changing it to a reference interfere with that?

Reply:
So I just tried to change the `newMap` vector to a reference and it indeed collides with `std::move` (the compilation fails). I would expect `std::move` to avoid 2 copies: one for the call, and another copy from the source to the new object, as the original source can be re-used. While passing it through a reference would only avoid one copy (in the call). Or am I missing something (@PhysSong)?

But since the `Keymap` constructor will be called very rarely, the difference in performance probably won't matter, and other than that I don't really know of any pros/cons that would make one option better than the other.

Reply:
It's usual to pass variables by reference(no copy) and then copy(no `std::move`) in copy constructors. Also, I suggested passing `QString`s by const references somewhere.

Reply:
I will rework it to use references, then. The `std::move`s were suggested to me by Dom in previous review, but unlike references I can't say I understand how `std::move` _really_ works, so I actually prefer using references. (The fact I forgot to use them is a clear bug on my side though, thanks for spotting that.)

Reply:
I should have replied to this long ago, but here we go now:

Passing by const reference here forces a copy of the vector to be made. A const reference makes sense for parameters that are only observed, but if the constructor needs to retain a copy of the parameter, there's no reason to restrict the ability to move arguments into the constructor.

Of course, one can provide an overload of the constructor that takes an rvalue reference, but the number of overloads required grows exponentially in the number of parameters whose types can be both copied and moved. Passing by value also supports both copying and moving, but in a single overload (albeit with an extra move required to get the argument into its final resting place in a member variable). Clang-tidy advocates for this latter approach as part of its modernisation suite (see [modernize-pass-by-value](https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html)). (The Core Guidelines actually disagree with Clang here, but I don't think their advice makes sense. The reasons aren't really relevant here, but I'll happily rant if invited.)

It's worth reading up on move semantics in C++ if you're not entirely familiar with it; it can offer performance advantages in situations where you'd traditionally copy, and in some cases copying isn't possible and you have to move anyway.

Reply:
Unresolving this so my previous comment is visible; feel free to resolve again if you want to leave this as is.

Reply:
Interesting, thanks for explaining. If @PhysSong does not object (since he suggested using references), I would have no problem going back to the `std::move` version, it makes sense to me. Performance probably does not matter here, but if we plan to start using the `std::move` approach in other places as well, it would save time during any future refactoring.

Reply:
Since I received no further input or objections either here or on Discord, I reverted the commit that replaced `std::move` by references.